clippy_lints/src/macro_use.rs

```diff
@@ -62,122 +65,144 @@ pub struct MacroUseImports {
-        for mac in self.collected.values() {
-            if paths_match(mac, use_path) {
-                return make_path(mac, use_path)
-            }
-        }
-        format!("{}::<macro name>", use_path)
-    }
-}
-
-fn paths_match(mac: &MacroRefData, use_path: &str) -> bool {
-    let segs = mac.path.split("::")
-        .filter(|s| *s != "")
-        .collect::<Vec<_>>();
+/// This is somewhat of a fallback for imports from `std::prelude` because they
+/// are not recognized by `LateLintPass::check_item` `lcx.tcx.item_children(id)`
+fn make_path(mac: &MacroRefData, use_path: &str) -> String {
+    let segs = mac.path.split("::").filter(|s| *s != "").collect::<Vec<_>>();
 
-    if segs.starts_with(&["std"]) {
-        return PRELUDE.iter().any(|m| segs.contains(m))
+    if segs.starts_with(&["std"]) && PRELUDE.iter().any(|m| segs.contains(m)) {
+        return format!("std::prelude::{} is imported by default, remove `use` statement", mac.name);
     }
-    
-    segs.starts_with(&use_path.split("::").collect::<Vec<_>>())
-}
 
-fn make_path(mac: &MacroRefData, use_path: &str) -> String {
     if use_path.split("::").count() == 1 {
         return format!("{}::{}", use_path, mac.name);
     }
 
-    let segs = mac.path.split("::")
-        .filter(|s| *s != "")
-        .collect::<Vec<_>>();
-    
-    if segs.starts_with(&["std"]) && PRELUDE.iter().any(|m| segs.contains(m)) {
-        return format!("std::prelude::{}", mac.name);
-    }
-    
     mac.path.clone()
 }
 
 impl_lint_pass!(MacroUseImports => [MACRO_USE_IMPORTS]);
 
-impl EarlyLintPass for MacroUseImports {
-    fn check_item(&mut self, ecx: &EarlyContext<'_>, item: &ast::Item) {
+impl<'l, 'txc> LateLintPass<'l, 'txc> for MacroUseImports {
+    fn check_item(&mut self, lcx: &LateContext<'_, '_>, item: &hir::Item<'_>) {
         if_chain! {
-            if ecx.sess.opts.edition == Edition::Edition2018;
-            if let ast::ItemKind::Use(use_tree) = &item.kind;
+            if lcx.sess().opts.edition == Edition::Edition2018;
+            if let hir::ItemKind::Use(path, _kind) = &item.kind;
             if let Some(mac_attr) = item
                 .attrs
                 .iter()
                 .find(|attr| attr.ident().map(|s| s.to_string()) == Some("macro_use".to_string()));
+            if let Res::Def(DefKind::Mod, id) = path.res;
             then {
-                let import_path = snippet(ecx, use_tree.span, "_");
-                let span = mac_attr.span.clone();
-                self.imports.push((import_path.to_string(), span));
+                for kid in lcx.tcx.item_children(id).iter() {
+                    if let Res::Def(DefKind::Macro(_mac_type), mac_id) = kid.res {
+                        let span = mac_attr.span.clone();
+                        println!("{:#?}", lcx.tcx.def_path_str(mac_id));
+                        self.imports.push((lcx.tcx.def_path_str(mac_id), span));
+                    }
+                }
```

Comment:
This seems to work exactly how we need it to, so far it ignores all the modules the macro they are defined in but not found in. I need to figure out how to import from another crate into my test crate so I can have the macro stick in a module.

Reply:
For macros from another crate you can use the `mini-macro` subcrate in the Clippy repo: https://github.com/rust-lang/rust-clippy/tree/master/mini-macro

This is also used in some other tests and includes a few proc-macros.